### ISSUE TITLE
[API Proposal]: Efficient EndPoint serialization

### ISSUE THREAD

**Issue by antonfirsov** [dotnet/runtime#78993]
### Background and motivation

Serializing general `EndPoint` instances into buffers/structures to be used with low-level interop code requires instantiating a `SocketAddress` and copying its bytes to a separate buffer manually. This leads to 2 extra allocations: `SocketAddress` and the underlying buffer. This is one of the main reasons our SendTo/ReceiveFrom methods allocate a lot, see discussion in #30797.

We can workaround this by special-casing `IPEndPoint` (#78591), but that would introduce complexity to the Socket codebase if implemented across all sync and async variants and doesn't work with arbitrary endpoints. We can also expose `SocketAddress.Buffer` (#78757), but IMO it's a partial solution, since it doesn't help with the allocations.

Instead working around our own inefficient abstractions, we should consider adding an alternative modern and efficient API to serialize arbitrary `EndPoint`-s into user-provided buffers. This would help simplifying and improving socket code while keeping it compatible with arbitrary endpoints.

### API Proposal

```csharp
namespace System.Net;

public abstract class EndPoint
{
    // Existing:
    // public virtual SocketAddress Serialize();
    // public virtual EndPoint Create(SocketAddress socketAddress);

    // Returns true if the EndPoint is serializable into a linear native structure
    public virtual bool TryGetSocketAddressSize(out int size);
    public virtual void Serialize(Span<byte> socketAddress);
    public virtual EndPoint Create(ReadOnlySpan<byte> socketAddress);
}
```

### API Usage

```csharp
if (endPoint.TryGetSocketAddressSize(out int size))
{
    Span<byte> socketAddress = size < 256 ? stackalloc byte[size] : new byte[size];
    endPoint.Serialize(socketAddress);
    
    // use socketAddress
}
```

### Alternative Designs

I can't think of a significantly different design that would also delegate the buffer lifecycle management to the user.

**Comment by fbrosseau** [comment]
What would the default Create(ROS) implementation be to keep backward compatibility? A call to Create(SocketAddress)?

**Comment by antonfirsov** [comment]
Yes, that would be the only way to keep things compatible. `TryGetSocketAddressSize(out int)` is more tricky, I updated the Risks section.

**Comment by ericstj** [comment]
Seems odd that Create looks like a factory method but it's not static.

**Comment by antonfirsov** [comment]
@ericstj it's matching the [existing overload](https://learn.microsoft.com/en-us/dotnet/api/system.net.endpoint.create?view=net-7.0). The proposed overload takes `ReadOnlySpan<byte>` instead of a `SocketAddress`.

The usage is unusual, but it's legacy design: an existing `EndPoint` instance is used to create new instances of the same subclass from native socket address buffer, see [`Socket.cs`](https://github.com/dotnet/runtime/blob/13a9a3ce67b3ab88cd6c0a975a47ed856e005d42/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1669). It meant to be invoked by low-level networking implementations, while third parties can implement EndPoints of arbitrary types by overriding `Create` and `Serialize`.

**Comment by antonfirsov** [comment]
Triage: depends on the fate of #30797, moving to future.
